### ISSUE TITLE
Custom regex that contains X but doesn't contain Y gives false positives

### ISSUE THREAD

**Issue by berkevren** [realm/SwiftLint#5061]
### New Issue Checklist

- [ ] Updated SwiftLint to the latest version
- [ ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

I'm trying to find any `guard let` statement that contains "=" and doesn't contain "try", ".", or "(".
My regex: `guard let.*=(?!.*try)[^.|(]*$`
Here are my tests: https://regex101.com/r/g08l4U/1

```
guard let view = view {
guard let malSubStatus = try asd
guard let view {
guard let view = view.prop {
guard let view = try view {
guard let view = view(some) {
guard let endOfEndDay = Calendar.current.date(byAdding: endOfDayComponents, to: startOfDay()) else {
guard let savedUsers, let suggestedUsers else {
```

Only the first line is a match, which is what I want.
When I run the swiftlint script via xcode build phases, these strings are marked as violations, even though they shouldn't violate this regex:

```
guard let endOfEndDay = Calendar.current.date(byAdding: endOfDayComponents, to: startOfDay()) else {
guard let endOfEndDay = Calendar.current.date(
guard let saved, let suggested else {
```

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint lint
```

### Environment

* SwiftLint version `0.50.3`
* Installation method used `CocoaPods`
* Paste your configuration file: 
```
guard_let:
    regex: guard let.*=(?!.*try)[^.|(|\[]*$
    message: "Guard let."
    severity: error
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
* Which Xcode version are you using (check `xcodebuild -version`)? 14.3
* Do you have a sample that shows the issue?

These trigger violations.
`guard let endOfEndDay = Calendar.current.date(`
`guard let saved, let suggested else {`

**Comment by SimplyDanny** [comment]
Try with `guard let.*=(?!.*try)[^.|\\(|\\[]*$`. Notice the escaped `(` and `[`.

**Comment by berkevren** [comment]
When I used the regex you suggested xcode doesn't tag anything, valid or invalid. Is it possible the lint doesn't accept this format, or escaping with double `\\`?

Here's my entire rule:
```
guard_let:
    regex: guard let.*=(?!.*try)[^.|\\(|\\[]*$
    severity: error
```

**Comment by SimplyDanny** [comment]
You are right. Single backslash is sufficient. Yet, something with the regex is not correct. Let's try to match all characters that are not preceded by the tokens you excluded. Does
```text
guard let.*=((?!try|\.|\(|\[).)*$
```
work for you? In my Playground, it only matches the first line `guard let view = view else {}` like you require.

**Comment by berkevren** [comment]
Using `guard let.*=((?!try|\.|\(|\[).)*$`
Are you not getting false positive for guard let's that are followed by `.`? 
`guard let endOfEndDay = Calendar.current.date(byAdding: endOfDayComponents, to: startOfDay()) else {`
case gives a false positive for me, and I think cases where there's a `.` after the `=` (or `(` following a `=` case).

Here's something that's similar and should immediately compile in a playground
`guard let endOfEndDay = Calendar.current.date(byAdding: .day, value: 1, to: Date.now) else {`

**Comment by SimplyDanny** [comment]
I tried on a Swift file with the content

```swift
guard let view = view else {}
guard let malSubStatus = try asd else {}
guard let view else {}
guard let view = view.prop else {}
guard let view = try view else {}
guard let view = view(some) else {}
guard let endOfEndDay = Calendar.current.date(byAdding: endOfDayComponents, to: startOfDay()) else {}
guard let savedUsers, let suggestedUsers else {}
guard let endOfEndDay = Calendar.current.date(byAdding: .day, value: 1, to: Date.now) else {}
```

and SwiftLint only reports the first line when using the configuration

```yaml
only_rules:
  - custom_rules

custom_rules:
  guard_let:
    regex: (?-s)guard let.*=(?!.*try)[^.|\(|\[]*$
    severity: error
```

The `(?-s)` seems to be important as well since SwiftLint would otherwise make `.` match newlines and then nothing is reported interestingly.

**Comment by berkevren** [comment]
This is the solution, thanks so much @SimplyDanny, closing the issue 😄
